In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Load the data

In [2]:
data = pd.read_csv(r"C:\Users\athir\data_sets\ML\Assg_community\placement.csv")
data.head()

cgpa  placement_exam_marks  placed
0  7.19                  26.0       1
1  7.46                  38.0       1
2  7.54                  40.0       1
3  6.42                   8.0       1
4  7.23                  17.0       0

In [3]:
data.shape

(1000, 3)

In [4]:
# step 3: data cleaning, data wrangling, data preprocessing

In [5]:
data.isnull().sum()

cgpa                    0
placement_exam_marks    0
placed                  0
dtype: int64

In [6]:
## Feature scaling

In [7]:
from sklearn.preprocessing import MinMaxScaler
mms = MinMaxScaler()
mms

MinMaxScaler()

In [8]:
data[['cgpa','placement_exam_marks']] = mms.fit_transform(data[['cgpa','placement_exam_marks']])
data.head()

cgpa  placement_exam_marks  placed
0  0.543735                  0.26       1
1  0.607565                  0.38       1
2  0.626478                  0.40       1
3  0.361702                  0.08       1
4  0.553191                  0.17       0

In [9]:
# separate x and y

In [10]:
x = data.drop('placed',axis=1)
y = data['placed']

In [11]:
# split the data

In [12]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size = 0.2, random_state = 0)

In [13]:
# apply KNeighborsClassifier on the train data

In [14]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier()
knc

KNeighborsClassifier()

In [15]:
knc.fit(x_train,y_train)

KNeighborsClassifier()

In [16]:
y_pred = knc.predict(x_test)
y_pred

array([1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0,
       1, 1], dtype=int64)

In [17]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.495

# Perform Hyperparameter tuning on the data to determine the combination of hyperparameters giving the highest accuracy.


In [18]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier()
parameters = {'n_neighbors':[3,5,7,9,11,13,15,17,19,21,23,25],'p':[1,2]}

In [19]:
from sklearn.model_selection import GridSearchCV
gscv = GridSearchCV(knc,parameters)
gscv

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23,
                                         25],
                         'p': [1, 2]})

In [20]:
gscv.fit(x_train,y_train)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23,
                                         25],
                         'p': [1, 2]})

In [21]:
gscv.cv_results_

{'mean_fit_time': array([0.00432196, 0.00338216, 0.        , 0.        , 0.        ,
        0.        , 0.00312405, 0.00312471, 0.        , 0.00312638,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.00314641, 0.        , 0.00310497,
        0.00014467, 0.0031271 , 0.        , 0.00312109]),
 'std_fit_time': array([0.00519242, 0.0061549 , 0.        , 0.        , 0.        ,
        0.        , 0.00624809, 0.00624943, 0.        , 0.00625277,
        0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.00629282, 0.        , 0.00620995,
        0.00028934, 0.0062542 , 0.        , 0.00624218]),
 'mean_score_time': array([0.00652847, 0.        , 0.0063199 , 0.00627236, 0.00625439,
        0.00312519, 0.00312452, 0.00316591, 0.00629606, 0.00327091,
        0.00613151, 0.00625248, 0.0062675 , 0.00624728, 0.00611219,
        0.00627651, 0.00746679, 0.00190287, 0.006252  , 0.        ,
        0.003002

In [22]:
res = pd.DataFrame(gscv.cv_results_)
res

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.004322      0.005192         0.006528        0.003247   
1        0.003382      0.006155         0.000000        0.000000   
2        0.000000      0.000000         0.006320        0.007741   
3        0.000000      0.000000         0.006272        0.007694   
4        0.000000      0.000000         0.006254        0.007660   
5        0.000000      0.000000         0.003125        0.006250   
6        0.003124      0.006248         0.003125        0.006249   
7        0.003125      0.006249         0.003166        0.006332   
8        0.000000      0.000000         0.006296        0.007711   
9        0.003126      0.006253         0.003271        0.006170   
10       0.000000      0.000000         0.006132        0.007512   
11       0.000000      0.000000         0.006252        0.007658   
12       0.000000      0.000000         0.006267        0.007676   
13       0.000000      0.000000         0.006247        0.007651   
14       0.000000      0.000000         0.006112        0.007488   
15       0.000000      0.000000         0.006277        0.007687   
16       0.000000      0.000000         0.007467        0.009360   
17       0.003146      0.006293         0.001903        0.003806   
18       0.000000      0.000000         0.006252        0.007657   
19       0.003105      0.006210         0.000000        0.000000   
20       0.000145      0.000289         0.003002        0.006004   
21       0.003127      0.006254         0.003128        0.006256   
22       0.000000      0.000000         0.003127        0.006254   
23       0.003121      0.006242         0.003128        0.006255   

   param_n_neighbors param_p                       params  split0_test_score  \
0                  3       1   {'n_neighbors': 3, 'p': 1}            0.53750   
1                  3       2   {'n_neighbors': 3, 'p': 2}            0.56250   
2                  5       1   {'n_neighbors': 5, 'p': 1}            0.53750   
3                  5       2   {'n_neighbors': 5, 'p': 2}            0.51250   
4                  7       1   {'n_neighbors': 7, 'p': 1}            0.52500   
5                  7       2   {'n_neighbors': 7, 'p': 2}            0.55000   
6                  9       1   {'n_neighbors': 9, 'p': 1}            0.51875   
7                  9       2   {'n_neighbors': 9, 'p': 2}            0.55000   
8                 11       1  {'n_neighbors': 11, 'p': 1}            0.54375   
9                 11       2  {'n_neighbors': 11, 'p': 2}            0.53125   
10                13       1  {'n_neighbors': 13, 'p': 1}            0.53125   
11                13       2  {'n_neighbors': 13, 'p': 2}            0.50625   
12                15       1  {'n_neighbors': 15, 'p': 1}            0.53125   
13                15       2  {'n_neighbors': 15, 'p': 2}            0.52500   
14                17       1  {'n_neighbors': 17, 'p': 1}            0.48750   
15                17       2  {'n_neighbors': 17, 'p': 2}            0.47500   
16                19       1  {'n_neighbors': 19, 'p': 1}            0.47500   
17                19       2  {'n_neighbors': 19, 'p': 2}            0.49375   
18                21       1  {'n_neighbors': 21, 'p': 1}            0.46875   
19                21       2  {'n_neighbors': 21, 'p': 2}            0.51250   
20                23       1  {'n_neighbors': 23, 'p': 1}            0.49375   
21                23       2  {'n_neighbors': 23, 'p': 2}            0.55625   
22                25       1  {'n_neighbors': 25, 'p': 1}            0.51250   
23                25       2  {'n_neighbors': 25, 'p': 2}            0.55625   

    split1_test_score  split2_test_score  split3_test_score  \
0             0.50625            0.54375            0.46250   
1             0.46875            0.52500            0.47500   
2             0.55000            0.61250            0.46875   
3             0.53125            0.59375          

In [23]:
res[['params','mean_test_score','rank_test_score']]

params  mean_test_score  rank_test_score
0    {'n_neighbors': 3, 'p': 1}          0.50875               15
1    {'n_neighbors': 3, 'p': 2}          0.51125               14
2    {'n_neighbors': 5, 'p': 1}          0.52875                4
3    {'n_neighbors': 5, 'p': 2}          0.51625               12
4    {'n_neighbors': 7, 'p': 1}          0.53500                1
5    {'n_neighbors': 7, 'p': 2}          0.52250                7
6    {'n_neighbors': 9, 'p': 1}          0.52375                6
7    {'n_neighbors': 9, 'p': 2}          0.53500                1
8   {'n_neighbors': 11, 'p': 1}          0.52250                8
9   {'n_neighbors': 11, 'p': 2}          0.52625                5
10  {'n_neighbors': 13, 'p': 1}          0.52000               11
11  {'n_neighbors': 13, 'p': 2}          0.53250                3
12  {'n_neighbors': 15, 'p': 1}          0.52250                8
13  {'n_neighbors': 15, 'p': 2}          0.52125               10
14  {'n_neighbors': 17, 'p': 1}          0.51250               13
15  {'n_neighbors': 17, 'p': 2}          0.49875               17
16  {'n_neighbors': 19, 'p': 1}          0.49000               21
17  {'n_neighbors': 19, 'p': 2}          0.49500               18
18  {'n_neighbors': 21, 'p': 1}          0.48375               24
19  {'n_neighbors': 21, 'p': 2}          0.50500               16
20  {'n_neighbors': 23, 'p': 1}          0.48625               23
21  {'n_neighbors': 23, 'p': 2}          0.49375               19
22  {'n_neighbors': 25, 'p': 1}          0.48750               22
23  {'n_neighbors': 25, 'p': 2}          0.49125               20

- Here the accuracy depending on the K value and the distance metric also.
- There are two combinations of hyperparameter giving the highest accuracy:
1. k value is 7 and Manhattan distance metric - 53.5% accuracy
2. k value is 9 and Euclidian distance metric - 53.5% accuracy.